src/crl.rs

```diff
@@ -732,12 +732,88 @@ impl<'a> IssuingDistributionPoint<'a> {
+    /// * General names of a type other than URI.
+    /// * Malformed names or invalid IDP or CRL DP extensions.
+    pub(crate) fn authoritative_for(&self, cert: &Cert<'a>) -> bool {
+        assert!(!self.only_contains_attribute_certs); // We check this at time of parse.
+
+        // Check that the scope of the CRL issuing distribution point could include the cert.
+        if self.only_contains_ca_certs && matches!(cert.ee_or_ca, EndEntityOrCa::EndEntity)
+            || self.only_contains_user_certs && matches!(cert.ee_or_ca, EndEntityOrCa::Ca(_))
+        {
+            return false;
+        }
+
+        let cert_dps = match cert.crl_distribution_points() {
+            // If the certificate has no distribution points, then the CRL can be authoritative
+            // based on the issuer matching and the scope including the cert.
+            None => return true,
+            Some(cert_dps) => cert_dps,
+        };
+
+        let mut idp_general_names = match self.names() {
+            Ok(Some(DistributionPointName::FullName(general_names))) => general_names,
+            _ => return false, // Note: Either no full names, or malformed. Shouldn't occur, we check at CRL parse time.
+        };
+
+        for cert_dp in cert_dps {
+            let cert_dp = match cert_dp {
+                Ok(cert_dp) => cert_dp,
+                // certificate CRL DP was invalid, can't match.
+                Err(_) => return false,
+            };
+
+            // If the certificate CRL DP was for an indirect CRL, or a CRL
+            // sharded by revocation reason, it can't match.
+            if cert_dp.crl_issuer.is_some() || cert_dp.reasons.is_some() {
+                return false;
+            }
+
+            let mut dp_general_names = match cert_dp.names() {
+                Ok(Some(DistributionPointName::FullName(general_names))) => general_names,
+                _ => return false, // Either no full names, or malformed.
+            };
+
+            // At least one URI type name in the IDP full names must match a URI type name in the
+            // DP full names.
+            if Self::uri_name_in_common(&mut idp_general_names, &mut dp_general_names) {
+                return true;
+            }
+        }
+
+        false
+    }
+
+    fn uri_name_in_common(
+        idp_general_names: &mut DerIterator<'a, GeneralName<'a>>,
+        dp_general_names: &mut DerIterator<'a, GeneralName<'a>>,
+    ) -> bool {
+        for name in idp_general_names.flatten() {
+            if let GeneralName::UniformResourceIdentifier(uri) = name {
+                for other_name in (&mut *dp_general_names).flatten() {
+                    if let GeneralName::UniformResourceIdentifier(other_uri) = other_name {
+                        if uri.as_slice_less_safe() == other_uri.as_slice_less_safe() {
+                            return true;
+                        }
+                    }
+                }
+            }
```

Comment:
Something like this?

```suggestion
            use GeneralName::UniformResourceIdentifier;
            let uri = match name {
                UniformResourceIdentifier(uri) => uri,
                _ => continue;
            };
            
            for other_name in (&mut *dp_general_names).flatten() {
                match other_name {
                    UniformResourceIdentifier(other_uri) if uri.as_slice_less_safe() == other_uri.as_slice_less_safe() => return true,
                    _ => continue,
                }
            }
```

Reply:
Ahh ok, that is clearer :+1: Thanks for the help. Updated.